### ISSUE TITLE
Cross-platform aggregation demo

### ISSUE THREAD

**Issue by SichangHe** [FedCampus/FedKit#25]
We now have worked around #18 and would allow for training the same model on both Android and iOS.

This demo would use MNIST because that is what the iOS example client app is using, and I anticipate it to be easier to port it to Android than the other way around.

- [ ] Use Flutter to handle downloading MNIST training data files, and make the iOS example work.
- [ ] Connect the Android example.

**Comment by SichangHe** [comment]
Flutter will download this:

[MNIST_data.zip](https://github.com/FedCampus/FedKit/files/12707552/MNIST_data.zip)

**Comment by SichangHe** [comment]
<details><summary>Picture</summary>

![16881695533621_ pic_hd](https://github.com/FedCampus/FedKit/assets/84777573/e41d2ff9-2a79-4a74-a5d7-7e0bdbcdaca3)

</details>

I got an Android and iOS phone training together, but the iPhone seems to be outputting bogus loss.

**Comment by SichangHe** [comment]
<details><summary>Picture</summary>

![16901695534208_ pic_hd](https://github.com/FedCampus/FedKit/assets/84777573/fc12fd43-a5a8-4f09-928b-3c46466798ae)

</details>

When I train with the two phone with the same partition ID (same training data), they converged in about two rounds.

**Comment by SichangHe** [comment]
<details><summary>Screenshot</summary>

![image](https://github.com/FedCampus/FedKit/assets/84777573/b97307e3-ceda-4e1b-94f0-2a7869c29945)

</details>

After changing the code to obtain `MLUpdateContext.metrics[.lossValue]` as `Float` instead of `Double`, I still get bogus losses.

@danielnugraha, any ideas 🙏?

**Comment by SichangHe** [comment]
I also suspect that we have communication encoding error between Swift and Flutter:

```elixir
Epoch 1 finished with loss 6557399552.000000
flutter: ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
flutter: │ #0   _MyAppState.appendLog (package:fed_kit_client/main.dart:69:12)
flutter: │ #1   _RootZone.runUnaryGuarded (dart:async/zone.dart:1594:10)
flutter: ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
flutter: │ 🐛 appendLog: Test accuracy after this round: -655739977728.0.
flutter: └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

**Comment by SichangHe** [comment]
We can exclude the possibility of communication error. I've tested and logged both of them out and they are the same.

**Comment by SichangHe** [comment]
Our data loading is wrong. I realized after reading this: https://github.com/JacopoMangiavacchi/MNIST-CoreML-Training/blob/02d57c3dcaf43420d33bad2c5740809fc76e84ba/MNIST-CoreML-Training/MNIST.swift#L187C1-L195C1.

**Comment by SichangHe** [comment]
This is the current model we are using, @luuvy757:

https://github.com/SichangHe/FedCampus--FedKit/blob/1bdeeed38a670944a4d134d318011b1287486f89/convert_model/mnist_eg/__init__.py#L21

**Comment by SichangHe** [comment]
<details><summary>Screenshot: 0 training loss, large test loss, and 0.1 test accuracy</summary>

![image](https://github.com/FedCampus/FedKit/assets/84777573/90b2c457-df04-44d5-a28e-5624cce9e4d3)

**Comment by SichangHe** [comment]
<details><summary>Screenshot of loss/accuracy.</summary>

![image](https://github.com/FedCampus/FedKit/assets/84777573/a67c626f-bf6d-4bb7-b202-1860cd59c6f6)

</details>

The binary classifier seems to give 100% accuracy out of the box on iOS and training did nothing.

Branch: https://github.com/SichangHe/FedCampus--FedKit/tree/android-mnist

Edit: The accuracy implementation on iOS is argmax so it is wrong.
However, training on Android yielded the same problem: the test loss&accuracy are the same over epochs (training loss did fluctuate).

Edit: Assigning random parameters did not work.

**Comment by SichangHe** [comment]
<details><summary>Screenshot</summary>

![image](https://github.com/FedCampus/FedKit/assets/84777573/fbb5b800-7077-4ebb-adf9-da014879a35b)

</details>

The data on iPhone is still crazy after switching to PMData.
Android result is fine, though.

**Comment by SichangHe** [comment]
<details><summary>Screenshot</summary>

![image](https://github.com/FedCampus/FedKit/assets/84777573/1962e852-a227-4d64-a088-0b19c2130831)

</details>

Even worse, I think the iPhone eventually managed to make all parameters NaN.
